python-avd/pyavd/_eos_designs/structured_config/network_services/utils.py

```diff
@@ -402,3 +402,57 @@ def get_vlan_aware_bundle_rt(
             return f"{admin_subfield}:{rt_override}"
 
         return f"{admin_subfield}:{bundle_number}"
+
+    def get_vrf_router_id(
+        self: AvdStructuredConfigNetworkServices,
+        vrf: EosDesigns._DynamicKeys.DynamicNetworkServicesItem.NetworkServicesItem.VrfsItem,
+        router_id: str,
+        tenant_name: str,
+    ) -> str | None:
+        """
+        Determine the router ID for a given VRF based on its configuration.
+
+        Args:
+            vrf (VrfsItem): The VRF object containing OSPF/BGP and vtep_diagnostic details.
+            router_id (str): The router ID type specified for the VRF (e.g., "vtep_diagnostic", "main_router_id", "none", or "IPv4 address").
+            tenant_name (str): The name of the tenant to which the VRF belongs.
+
+        Returns:
+            str | None: The resolved router ID as a string, or None if the router ID is not applicable.
+
+        Raises:
+            AristaAvdInvalidInputsError: If required configuration for "vtep_diagnostic" router ID is missing.
+        """
+        # Handle "vtep_diagnostic" router ID case
+        if router_id == "vtep_diagnostic":
+            vrf_diagnostic_loopback = vrf.vtep_diagnostic.loopback
```

Comment:
I think we can use, self._get_vtep_diagnostic_loopback_for_vrf(vrf) here to get the loopback ip.

Reply:
Used the mention method but after validating the error. Also added negative unit tests

Reply:
We can raise the error below the mentioned method if it returns none. Error should be loopback and loopback_ip_range or loopback_ip_pool is required.

Reply:
updated